client/checkout/blocks/saved-token-handler.js

```diff
@@ -1,15 +1,41 @@
 /**
  * Internal dependencies
  */
+import { useEffect } from 'react';
 import { usePaymentCompleteHandler } from './hooks';
+import { useSelect } from '@wordpress/data';
 
 export const SavedTokenHandler = ( {
 	api,
 	stripe,
 	elements,
-	eventRegistration: { onCheckoutSuccess },
+	eventRegistration: { onPaymentSetup, onCheckoutSuccess },
 	emitResponse,
 } ) => {
+	const paymentMethodData = useSelect( ( select ) => {
+		const store = select( 'wc/store/payment' );
+		return store.getPaymentMethodData();
+	} );
+
+	useEffect( () => {
+		onPaymentSetup( async () => {
```

Comment:
You need to `return` the result of `onPaymentSetup()` because this returns an "unsubscribe" function. When this `useEffect` reruns, it'll execute the return value of the last run (if it's a function), therefore unsubscribing the old listener before resubscribing the new one.

Currently this will add a new listener to `onPaymentSetup` every time the payment method data changes.

Reply:
TIL. Good thinking.

Reply:
@FangedParakeet @bborman22 @kalessil shall I go on and add `return` to the other instances too?

Reply:
Addressed in [6dead9c](https://github.com/Automattic/woocommerce-payments/pull/7701/commits/6dead9cb960647545e583cd8e472f8e8882bf02a), thanks for the review!

Reply:
As an additional test to make sure it's working, I'd recommend:

- add a console log to the `onPaymentSetup` function
- manually set the payment method data multiple times (`wp.data.dispatch( 'wc/store/payment').__internalSetPaymentMethodData( { test: true} )`)
- Reset the payment method data to what you want it to be
- Try to pay and ensure only one console log is output.

Reply:
Well, it doesn't work as it's in the `SavedTokenHandler`, and it's only running just before the payment starts processing. Actually, to test it, I need to set a declined card as a saved payment method, giving an error like "Insufficient funds", or something.

Reply:
Ah ok, well I just think it'd be useful to check it's unsubscribed when `paymentMethodData` changes to prevent a potential memory leak. Code looks fine as it is but it would be nice to double check!